### ISSUE TITLE
Gemini detect difference in values for column with list type. 

### ISSUE THREAD

**Issue by aleksbykov** [scylladb/scylladb#7937]
*Installation details*
Scylla version (or git commit hash): Scylla version 4.2.3-0.20210111.24346215c
Cluster size: 3
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-0e14ad8f72d28f53a(eu-north-1)
Scylla oracle version: 3.3.2

During this job: https://jenkins.scylladb.com/job/scylla-4.2/job/gemini-/job/gemini-3h-with-nemesis-test/94/, gemini failed during MultiHardReboot nemesis with next error:
```
col5:[2003-06-15 00:00:00 +0000 UTC 2001-03-24 00:00:00 +0000 UTC 1991-10-28 00:00:00 +0000 UTC 2001-01-05 00:00:00 +0000 UTC 1981-08-07 00:00:00 +0000 UTC 2003-06-15 00:00:00 +0000 UTC 2001-03-24 00:00:00 +0000 UTC 1991-10-28 00:00:00 +0000 UTC 2001-01-05 00:00:00 +0000 UTC 1981-08-07 00:00:00 +0000 UTC] col6:eb6b1580-a1fc-11c3-a5cf-06cadd9c1c54 col7:0.8803340122776178 col8:928808429318907682 col9:Rd0FXdRt4gCONA4tmqyOAEXUYlS9yiEzrr/1OWJwak1Sa9ERyUX88qmLkfCN81sw/JlyoFh/oQiLY2ucNJ33HLx5nIQ1Fwznc99gyWV2YuqqbdTJhfffNxfmsNxjNYzxvS0XatCmL80b5RMUxqBQWETdtU5Do/NLR2+UhQSgxnGj2PTH4reF2b3gZpqINlpjF9k0i/X6vKwvjAip5dLAqmjuz5mJ7Rmu17MGNfWVUZaflVZk0iwlwC4Z3C/YVPdu3LF1O6TrgXVu1jKS9yxuRLRBVgWp2LZDD2DybAFBp+N7QIZ5isoCF7gtjAdgLc6sH2qL/opmG7W9t0XhGOCDy3Hd0d7ZkRGK38NSriL99fg9ydrAcGFS9PfFh7cX3QrD5BMDtY7rvjWQ9BhzKX3DnjGd6vqgRVINhwmBMdlz6rqdMAYf9Ara++DQLK84I7AMFioBvyH3ZkbB6gFwWoQl1M+0VGDiF1q88AP9D2OdQ84gi3aojJRCnqQxg3bdvZuspKX7+MWx09xZLlqN/Q1O5VE13Hra7m+QMY6WP6Wk1U1NBkSK5pRj+pgawg2SIqbyuK5eIqPTRFJ5X/9Fu3sG3RJRLtvqTWz3swRUVWkHxmsC3Qj+M04qj+EGnkis0IoKSE0CERzmzOQAeDAulDvXGXZg8powhd7vIpsh8YGN9o5qEPB7vCX1A8E5xz/a7Hj1xoHRbQ6gZFYrii4vCHKOHT1VibFYLCXfG7K7OPjW6C3ENW9GjgLZx1JbN2gZFPiqB+6e7iAgVH8YYYsnETfsXYXY6DolQIjZ1IJYQjienv9ns+ev4sDvfhk24PVVehzyjsAa28wlHuZQNRjobDo9MQ3JhPZc00/w1BBw7E0fbJu2cZgfUlfVd9OG+cS0N/7qi9RChkK2And85gFGR3rI5lfj/t6BzlSDqcBni9xp pk0:35 pk1:2117705902 pk2:1003189364 pk3:8567842351728392330 pk4:1915980338]):   map[string]interface{}{\n  \t... // 9 identical entries\n  \t\"col3\": string(\"VKGsFKvF31fVeW3A5y8MryRmNZyPq01ADw2D8OSs6/R7oDrUxJccX9nYTf0vnkscwz9FOFlZTxcEVRf1fJzP/FfviH8NBGxQxTCteocunxxp/nPF5SF+1lAEt8FrGHk4g8Sb75+SaTheLQPqqG95V3DIp+BDF9GpmmUsq7qQAhLNwoc/dJrHPD6EmkmgxYs+rdMUgOd4dChj5n8CxoZ2SAXuyb6BP9fd2YDdsMag18tlAZSdW7F/jS09S6MxoO690PRO+JpxEiX5ss/v678xWiiWowDgiA8SRAa7CwDoH5lqENm+WNtQV\"),\n  \t\"col4\": s\"4186685395014475.547\",\n  \t\"col5\": []time.Time{\n  \t\t... // 3 identical elements\n  \t\ts\"2001-01-05 00:00:00 +0000 UTC\",\n  \t\ts\"1981-08-07 00:00:00 +0000 UTC\",\n+ \t\ts\"2003-06-15 00:00:00 +0000 UTC\",\n+ \t\ts\"2001-03-24 00:00:00 +0000 UTC\",\n+ \t\ts\"1991-10-28 00:00:00 +0000 UTC\",\n+ \t\ts\"2001-01-05 00:00:00 +0000 UTC\",\n+ \t\ts\"1981-08-07 00:00:00 +0000 UTC\",\n  \t},\n  \t\"col6\": gocql.UUID{0xeb, 0x6b, 0x15, 0x80, 0xa1, 0xfc, 0x11, 0xc3, 0xa5, 0xcf, 0x06, 0xca, 0xdd, 0x9c, 0x1c, 0x54},\n  \t\"col7\": float64(0.8803340122776178),\n  \t... // 7 identical entries\n  }\n", Query:"SELECT * FROM ks1.table1 WHERE pk0=35 AND pk1=2117705902 AND pk2=1003189364 AND pk3=8567842351728392330 AND pk4=1915980338 "}}


col5:[2003-06-15 00:00:00 +0000 UTC 2001-03-24 00:00:00 +0000 UTC 1991-10-28 00:00:00 +0000 UTC 2001-01-05 00:00:00 +0000 UTC 1981-08-07 00:00:00 +0000 UTC]
col5:[2003-06-15 00:00:00 +0000 UTC 2001-03-24 00:00:00 +0000 UTC 1991-10-28 00:00:00 +0000 UTC 2001-01-05 00:00:00 +0000 UTC 1981-08-07 00:00:00 +0000 UTC 2003-06-15 00:00:00 +0000 UTC 2001-03-24 00:00:00 +0000 UTC 1991-10-28 00:00:00 +0000 UTC 2001-01-05 00:00:00 +0000 UTC 1981-08-07 00:00:00 +0000 UTC]
```

after that i reran the job with same seed: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/abykov/job/BYO-abykov-staging/60, iit again failed but now during DecommissionNemeis. After node3 was decommissioned and during node4 was being added to test cluster, gemini detected next error and failed:
```
col1:220905678 col10:82.33.201.159 col11:false col12:map[udt_2401624638_0:UAjZHbKGCoMoheaGZO6y4DrPAZWLUV9A6p3pzKlrxQXd5kCXqYd5tkhxmoaPWWisuuFdF+fK74S/atJcPm9rTNIKmJob8EK+6MOE68XcWpqUiywSR20hk07Rd6kymBvqXtIJiYz+HJZA7YcZyDdcSNCEnAXsBKIiRhgZpIyXn2BoCiSDl127jebIULCPEfIw4HePaJzm1DVib6JlHCyGZegEVyQxJwnv7yuhIsWuEoYxabURdY7yJwdYOBSSiGiCpU3NWwu+3pITqv2c2/SFIxJjJYn/TqiOxgyTpsV9DpKQwx6aXzRIlsYyRAa8tsrfESNxXeb2OSMTcItONnHNood6ZJ3xCObAQSnOwUh09sA7hCn0Wxn0Tt8Zwngy4S4wqMrzLmTlMmMvrVE1MPlYuP/sN/MR8dbgVUhyixOHG/XogiMYAxjSgR9BCbw72r+IhKotiYmuJj+vS06M3ZEl197EK+1ZBh4CXlYoSdQOEaYEQ8jMaA1jbNm5sMc8mA4/eJtO3xjBNjJXFuxSPhwRSol1URBP6BNHdaNH+HdpbyigMNIAAmzCB5XqGs1iwrAQJsxSo27GO6T24gLROJSy48YytJe78hky/sQ97q8AD47DYLA/kDf0NLtGhzYiqszBd2So0+DIntn udt_2401624638_1:00s9JhZVxcT2mYSdvMdfQio0AceJgb2YEEC+zIUfRusqbbaW1H835iA2Kp2/aQ2Y2QiO795Xe9uC9ygc3T3eqOZIc5ZTZATVdT4cF+L+S3VE+YbU1npY+zHYy8n5QNvbOvhbhL5F8wZwtJrXzABT8JM9kki3ncwXnG9nWcpKJS8C2QCjhUtjzZdaFxZUJ3UKsOV udt_2401624638_2:true udt_2401624638_3:Xmec01VOpj] col2:{0 0 3960000000000} col3:RKjVblgwC6e3U58JV8hLn1CIm2Ic+J99QX433070oNsarABAbs0NZeC1rYjWbH1e5GGLW9IT2EpDSPAeZpBzLDUGQ9W483ELEFrAY86bjdAuDpCZPmfWguZVp9nRhwfndESjCuBDTKJ68ccL9t2vwK1BYVS6m/TB6HTgdWcRKhqHKreTddEspYPNr/eZrxCkKp6/bp6Nre90EPvw+CatX9+C3H445EPVS+9zX/ibh3lke+1GCA col4:5628445346040249.954 col5:[1970-05-27 00:00:00 +0000 UTC 2020-08-27 00:00:00 +0000 UTC 2007-11-10 00:00:00 +0000 UTC 1985-01-26 00:00:00 +0000 UTC 1980-06-24 00:00:00 +0000 UTC 1970-05-27 00:00:00 +0000 UTC 2020-08-27 00:00:00 +0000 UTC 2007-11-10 00:00:00 +0000 UTC 1985-01-26 00:00:00 +0000 UTC 1980-06-24 00:00:00 +0000 UTC] col6:7e990680-c7eb-11e8-ba3f-06aa7c6558c8 col7:0.6736832508112683 col8:3130967255677260920 col9:Z95w18f7KIP2XrLOoQ3DwJChUMspvTuI/O4qUTHAgRYSpn4vKiFgmCQNcsYntZKHIS2Cdh7TZrVi3c2BcxIuDkH2454xavXhjuMRr62sWr7CNgqRIOy2k5g6A1jysg7lXJZnccUs1aEE+NyBU1s+7sIA1etNH/DVuDqgZPFs9n3jyD6avb1LDXtfBOwolBNyEX/hp8m+2HE27HIKjTfLEc61ACWAhzEpzMAH7tEN52W8tjrVfG+h4j49R1dk3A9wb5fQvwJsDzGNVHHV/+ztLx91CaGo2zpygRcfxOgEPQBP04gjchqvutrzCi6x2Ys9ucy4n1swnhBUJ+nIO4fJs2x+p5xHHhUc7q2ZipNmGA5xx2NeW1pYk7XvsULKPIOcoLR8HFMEXwEtm816NnYSHmoqo+SrQ5+4/0Zm0UyyuivDADaEDbwK4Zk0XpeffQIcMNZoINLynKmJ81rf+6J8VKTA35QST/GprrBnwrkkyiqecbQRTeTItI5fIFTo+30gciafa8LK86VPtWH pk0:-17 pk1:604504399 pk2:1732162752 pk3:8209381498373038186 pk4:270920682]): \u00a0\u00a0map[string]interface{}{\n\u00a0\u00a0\t... // 9 identical entries\n\u00a0\u00a0\t\"col3\": string(\"RKjVblgwC6e3U58JV8hLn1CIm2Ic+J99QX433070oNsarABAbs0NZeC1rYjWbH1e5GGLW9IT2EpDSPAeZpBzLDUGQ9W483ELEFrAY86bjdAuDpCZPmfWguZVp9nRhwfndESjCuBDTKJ68ccL9t2vwK1BYVS6m/TB6HTgdWcRKhqHKreTddEspYPNr/eZrxCkKp6/bp6Nre90EPvw+CatX9+C3H445EPVS+9zX/ibh3lke+1GCA\"),\n\u00a0\u00a0\t\"col4\": s\"5628445346040249.954\",\n\u00a0\u00a0\t\"col5\": []time.Time{\n\u00a0\u00a0\t\t... // 3 identical elements\n\u00a0\u00a0\t\ts\"1985-01-26 00:00:00 +0000 UTC\",\n\u00a0\u00a0\t\ts\"1980-06-24 00:00:00 +0000 UTC\",\n+\u00a0\t\ts\"1970-05-27 00:00:00 +0000 UTC\",\n+\u00a0\t\ts\"2020-08-27 00:00:00 +0000 UTC\",\n+\u00a0\t\ts\"2007-11-10 00:00:00 +0000 UTC\",\n+\u00a0\t\ts\"1985-01-26 00:00:00 +0000 UTC\",\n+\u00a0\t\ts\"1980-06-24 00:00:00 +0000 UTC\",\n\u00a0\u00a0\t},\n\u00a0\u00a0\t\"col6\": gocql.UUID{0x7e, 0x99, 0x06, 0x80, 0xc7, 0xeb, 0x11, 0xe8, 0xba, 0x3f, 0x06, 0xaa, 0x7c, 0x65, 0x58, 0xc8},\n\u00a0\u00a0\t\"col7\": float64(0.6736832508112683),\n\u00a0\u00a0\t... // 7 identical entries\n\u00a0\u00a0}\n", Query:"SELECT * FROM ks1.table1 WHERE pk0=-17 AND pk1=604504399 AND pk2=1732162752 AND pk3=8209381498373038186 AND pk4=270920682 "}}
```

the query and failed column is the same.

After that i stopped scylla on each node in test cluster. Next i enabled scylla on the node and run failed query directly from python script to test node and oracle node. I repeated these steps for each node and each time query showed difference in same column for each test node versus oracle node:
```
ame pk0
same pk1
same pk2
same pk3
same pk4
same ck0
same ck1
same ck2
same col0
same col1
same col10
same col11
same col12
same col2
same col3
same col4
row error found. Column: col5
same col6
same col7
same col8
same col9
```

```
:[AssertionError("field:col5 are differs:
t_row:[Date(146), Date(18501), Date(13827), Date(5504), Date(3827), Date(146), Date(18501), Date(13827), Date(5504), Date(3827)] on test cluster
o_row:[Date(146), Date(18501), Date(13827), Date(5504), Date(3827)] on oracle cluster.
```
Values of column col5 on all test nodes are the same and looks like duplicated

Nodes are still alive:
```
+---------------------------------------------------------------+-------------+---------------+
| SCT-Runner                                                    | eu-north-1a | 13.49.74.120  |
| gemini-with-nemesis-3h-normal-maste-db-node-e1cf0ff7-1        | eu-north-1a | 13.53.127.228 |
| gemini-with-nemesis-3h-normal-maste-db-node-e1cf0ff7-2        | eu-north-1a | 13.49.245.203 |
| gemini-with-nemesis-3h-normal-maste-oracle-db-node-e1cf0ff7-1 | eu-north-1a | 13.51.48.75   |
| gemini-with-nemesis-3h-normal-maste-monitor-node-e1cf0ff7-1   | eu-north-1a | 13.49.72.113  |
| gemini-with-nemesis-3h-normal-maste-db-node-e1cf0ff7-4        | eu-north-1a | 13.53.125.217 |
+---------------------------------------------------------------+-------------+---------------+
```
to start scylla on the single node, i manually update scylla.yaml seed nodes with current node ip.

Could you help where to look further and next steps?

this is the schema generated by gemini:
```
cqlsh> desc KEYSPACE ks1 ;

CREATE KEYSPACE ks1 WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;

CREATE TYPE ks1.udt_2401624638 (
    udt_2401624638_0 ascii,
    udt_2401624638_1 ascii,
    udt_2401624638_2 boolean,
    udt_2401624638_3 text
);

CREATE TABLE ks1.table1 (
    pk0 tinyint,
    pk1 int,
    pk2 int,
    pk3 varint,
    pk4 int,
    ck0 inet,
    ck1 inet,
    ck2 bigint,
    col0 set<blob>,
    col1 int,
    col10 inet,
    col11 boolean,
    col12 frozen<udt_2401624638>,
    col2 duration,
    col3 text,
    col4 decimal,
    col5 list<date>,
    col6 uuid,
    col7 double,
    col8 bigint,
    col9 text,
    PRIMARY KEY ((pk0, pk1, pk2, pk3, pk4), ck0, ck1, ck2)
) WITH CLUSTERING ORDER BY (ck0 ASC, ck1 ASC, ck2 ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
CREATE INDEX table1_col1_idx ON ks1.table1 (col1);
CREATE INDEX table1_col7_idx ON ks1.table1 (col7);
CREATE INDEX table1_col4_idx ON ks1.table1 (col4);

CREATE MATERIALIZED VIEW ks1.table1_col7_idx_index AS
    SELECT col7, idx_token, pk0, pk1, pk2, pk3, pk4, ck0, ck1, ck2
    FROM ks1.table1
    WHERE col7 IS NOT NULL
    PRIMARY KEY (col7, idx_token, pk0, pk1, pk2, pk3, pk4, ck0, ck1, ck2)
    WITH CLUSTERING ORDER BY (idx_token ASC, pk0 ASC, pk1 ASC, pk2 ASC, pk3 ASC, pk4 ASC, ck0 ASC, ck1 ASC, ck2 ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE MATERIALIZED VIEW ks1.table1_col1_idx_index AS
    SELECT col1, idx_token, pk0, pk1, pk2, pk3, pk4, ck0, ck1, ck2
    FROM ks1.table1
    WHERE col1 IS NOT NULL
    PRIMARY KEY (col1, idx_token, pk0, pk1, pk2, pk3, pk4, ck0, ck1, ck2)
    WITH CLUSTERING ORDER BY (idx_token ASC, pk0 ASC, pk1 ASC, pk2 ASC, pk3 ASC, pk4 ASC, ck0 ASC, ck1 ASC, ck2 ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE MATERIALIZED VIEW ks1.table1_col4_idx_index AS
    SELECT col4, idx_token, pk0, pk1, pk2, pk3, pk4, ck0, ck1, ck2
    FROM ks1.table1
    WHERE col4 IS NOT NULL
    PRIMARY KEY (col4, idx_token, pk0, pk1, pk2, pk3, pk4, ck0, ck1, ck2)
    WITH CLUSTERING ORDER BY (idx_token ASC, pk0 ASC, pk1 ASC, pk2 ASC, pk3 ASC, pk4 ASC, ck0 ASC, ck1 ASC, ck2 ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE MATERIALIZED VIEW ks1.table1_mv_0 AS
    SELECT *
    FROM ks1.table1
    WHERE col7 IS NOT null AND pk0 IS NOT null AND pk1 IS NOT null AND pk2 IS NOT null AND pk3 IS NOT null AND pk4 IS NOT null AND ck0 IS NOT null AND ck1 IS NOT null AND ck2 IS NOT null
    PRIMARY KEY ((col7, pk0, pk1, pk2, pk3, pk4), ck0, ck1, ck2)
    WITH CLUSTERING ORDER BY (ck0 ASC, ck1 ASC, ck2 ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
```

Gemini command: 
```
gemini -d --duration 3h --warmup 30m -c 100 -m mixed -f --non-interactive --cql-features normal --max-mutation-retries 5 --max-mutation-retries-backoff 500ms --async-objects-stabilization-attempts 5 --async-objects-stabilization-backoff 500ms --replication-strategy "{'class': 'SimpleStrategy', 'replication_factor': '3'}" --oracle-replication-strategy "{'class': 'SimpleStrategy', 'replication_factor': '1'}" --test-cluster=10.0.0.133,10.0.2.182,10.0.3.22 --outfile /home/centos/gemini_result_b8346ef2-e030-4ef4-a2a7-5b86ddb4b253.log --seed 80 --oracle-cluster=10.0.1.47
```

db_logs for job  https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/abykov/job/BYO-abykov-staging/60: https://cloudius-jenkins-test.s3.amazonaws.com/e1cf0ff7-a165-4039-ae58-17c4c62d5202/20210119_094344/db-cluster-e1cf0ff7.zip

**Comment by aleksbykov** [comment]
Absolutely the same picture with test cluster is 4.3 and oracle cluster 4.1.10

**Comment by fgelcer** [comment]
experiencing very similar issue, i'm running `Scylla version 4.5.dev-0.20210124.020da49c89b with build-id 6087ff4830a3bd8ea78dec39851e3bdac0251e7f`

after the test failed, i re-run it with the same seed, and it failed once again, with the same query:
`SELECT * FROM ks1.table1 WHERE pk0=102 AND pk1=-14369 AND pk2=11113 ;`

what i noted is that the oracle's output is much smaller than the other nodes (all the test nodes are exactly the same result):
```
$ ls -ltrah
total 3.3M
-rw-rw-r-- 1 fabio fabio 870K Jan 27 12:55 node1_query_res.txt
-rw-rw-r-- 1 fabio fabio 870K Jan 27 12:56 node2_query_res.txt
-rw-rw-r-- 1 fabio fabio 870K Jan 27 12:57 node3_query_res.txt
-rw-rw-r-- 1 fabio fabio 629K Jan 27 12:58 oracle_query_res.txt
```

and comparing the test nodes:
```
$ diff node1_query_res.txt node2_query_res.txt
$ diff node1_query_res.txt node3_query_res.txt
```

here are the query results:
[node1_query_res.txt](https://github.com/scylladb/scylla/files/5880975/node1_query_res.txt)
[node2_query_res.txt](https://github.com/scylladb/scylla/files/5880976/node2_query_res.txt)
[node3_query_res.txt](https://github.com/scylladb/scylla/files/5880977/node3_query_res.txt)
[oracle_query_res.txt](https://github.com/scylladb/scylla/files/5880978/oracle_query_res.txt)


full test logs are here:
```
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|                                                                                                           Log links for testrun with test id 3c3bac9d-9226-4dcb-86dd-367eb9f0a8c2                                                                                                           |
+-----------------+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Date            | Log type    | Link                                                                                                                                                                                                                                                        |
+-----------------+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 20210126_153943 | grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/3c3bac9d-9226-4dcb-86dd-367eb9f0a8c2/20210126_153943/grafana-screenshot-byo-abykov-staging-scylla-per-server-metrics-nemesis-20210126_154329-gemini-with-nemesis-3h-normal-maste-monitor-node-3c3bac9d-1.png |
| 20210126_153943 | grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/3c3bac9d-9226-4dcb-86dd-367eb9f0a8c2/20210126_153943/grafana-screenshot-overview-20210126_153943-gemini-with-nemesis-3h-normal-maste-monitor-node-3c3bac9d-1.png                                             |
| 20210126_154816 | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/3c3bac9d-9226-4dcb-86dd-367eb9f0a8c2/20210126_154816/db-cluster-3c3bac9d.zip                                                                                                                                 |
| 20210126_154816 | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/3c3bac9d-9226-4dcb-86dd-367eb9f0a8c2/20210126_154816/loader-set-3c3bac9d.zip                                                                                                                                 |
| 20210126_154816 | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/3c3bac9d-9226-4dcb-86dd-367eb9f0a8c2/20210126_154816/monitor-set-3c3bac9d.zip                                                                                                                                |
| 20210126_154816 | sct-runner  | https://cloudius-jenkins-test.s3.amazonaws.com/3c3bac9d-9226-4dcb-86dd-367eb9f0a8c2/20210126_154816/sct-runner-3c3bac9d.zip                                                                                                                                 |
+-----------------+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

**Comment by fgelcer** [comment]
@aleksbykov gently confirmed that above i'm reporting the same issue. From a chat with Alex:
```
Exactly the same issue
only difference that list has blob elements:
len(test_col_value) == len(oracle_col_value)
check difference in lists
check the list length
Traceback (most recent call last):
 File "/home/abykov/Projects/scylla-cluster-tests/check_gemini_row.py", line 50, in <module>
   assert test_col_value == oracle_col_value
AssertionError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
 File "/home/abykov/Projects/scylla-cluster-tests/check_gemini_row.py", line 57, in <module>
   assert len(test_col_value) == len(oracle_col_value), f"len of list in {col_name} are different, looks like data duplicated\n test: {len(test_col_value)}: \n oracle: {len(oracle_col_value)}"
AssertionError: len of list in col6 are different, looks like data duplicated
test: 8:
oracle: 4
[Finished in 13.7s with exit code 1]
```

**Comment by slivne** [comment]
@eliransin List issue - please assign someone to look at it.

**Comment by slivne** [comment]
@eliransin ping

**Comment by eliransin** [comment]
@psarna please have a look. @avikivity mentioned that it might be an expected (bad) behavior of the list datatype.

**Comment by psarna** [comment]
Possibly related: https://github.com/scylladb/scylla/issues/3559

**Comment by psarna** [comment]
Kamil pointed me to a line of code in gemini which retries the queries:
https://github.com/scylladb/gemini/blob/e0014702159a4a787675a7caeb6d9265a8d9b16a/store/cqlstore.go#L49
, depending on the `--max-mutation-retries` parameter. Does anybody know what's the default value for it? In order to confirm that it's the root cause, it would be nice to log all retries - then we could check if a retried query could have caused the problem.

The issue with lists is that when an update with identical timestamp ends up on different nodes, it breaks list's idempotence.
Here's how to observe the issue with a local 3-node cluster:
1. Create a table:
```cql
CREATE KEYSPACE ks WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1} ; use ks;
CREATE TABLE list ( year int, month int, events list<TEXT>, PRIMARY KEY ( year, month) );
```
2. Issue an update from a **single** cqlsh session connected to just one node to see that it theoretically looks fine:
```cql
cassandra@cqlsh:ks> INSERT INTO list (year, month, events) VALUES (2015, 06, ['a','b','c']) using timestamp 1458019725702 ;
cassandra@cqlsh:ks> SELECT * FROM list;

 year | month | events
------+-------+-----------------
 2015 |     6 | ['a', 'b', 'c']

(1 rows)
```
3. Issue a simultaneous update from **three** cqlsh processes, each connected to a different node - it's going to be a race, but sending all three from a multiplexed tmux/screen/terminator session with a single click is enough to reproduce the problem:
```cql
-- execute the step below simultaneously on 3 cqlsh sessions, each connected to a different node
cassandra@cqlsh:ks> INSERT INTO list (year, month, events) VALUES (2015, 06, ['a','b','c']) using timestamp 1458019725702 ; 
cassandra@cqlsh:ks> SELECT * FROM list;

 year | month | events
------+-------+-----------------------------------------------
 2015 |     6 | ['a', 'a', 'a', 'b', 'b', 'b', 'c', 'c', 'c']

(1 rows)
```

So, perhaps this is a "feature" of lists, not a bug, as described in https://github.com/scylladb/scylla/issues/3559 . Adding more logs to gemini could help confirm the issue.

**Comment by psarna** [comment]
I was able to reproduce the failure locally with gemini and two local clusters. Here's the outcome from both:
1:
```cql
cassandra@cqlsh> SELECT writetime(col4), col5 FROM ks1.table1 WHERE pk0=-112 AND pk1=1813652260 AND pk2=648564855 AND pk3=5234052560646235106 AND pk4=728627908;

 writetime(col4)  | col5
------------------+--------------------------------------------------------------------------
 1619617314707858 | [1994-07-23, 2001-12-22, 2021-11-18, 1986-12-30, 2022-09-24, 1998-11-10]

(1 rows)
```

2:
```cql
cassandra@cqlsh> SELECT writetime(col4), col5 FROM ks1.table1 WHERE pk0=-112 AND pk1=1813652260 AND pk2=648564855 AND pk3=5234052560646235106 AND pk4=728627908;

 writetime(col4)  | col5
------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
 1619617314707858 | [1994-07-23, 2001-12-22, 2021-11-18, 1986-12-30, 2022-09-24, 1998-11-10, 1994-07-23, 2001-12-22, 2021-11-18, 1986-12-30, 2022-09-24, 1998-11-10]

(1 rows)
```

As you can see, the entries in cluster 2 are duplicated. It's unfortunately not possible to retrieve the writetime for unfrozen collection items, so I took the writetime value from another column, **but** the gemini logs suggest that all inserts/updates were always adding the whole row with all column values to the database. Given that, it can be reasonably assumed that all values of the collection were inserted with the same timestamp as well.

I also saw this timestamp (`1619617314707858`) in debug logs I added locally each time a request was retried from the client side, which means that an update with this exact timestamp was retried at least once. All that fits into #3559 - retrying with an identical timestamp on two nodes can lead to duplicated list values. Unfortunately, I didn't find an easy way to track which node was picked by the gocqlx driver each time it wanted to retry a request.

Please let me know if you want me to investigate further or if the current output convincing enough. I could also enable trace level logging on all nodes in both clusters and wait for this failure to happen again, but tracing all requests significantly slows down nodes, so it might take a long while, especially that this particular failure also reproduces only after 1-3 hours of uptime without elevated logging levels.

**Comment by aleksbykov** [comment]
> Kamil pointed me to a line of code in gemini which retries the queries:
> https://github.com/scylladb/gemini/blob/e0014702159a4a787675a7caeb6d9265a8d9b16a/store/cqlstore.go#L49
> , depending on the `--max-mutation-retries` parameter. Does anybody know what's the default value for it? In order to confirm that it's the root cause, it would be nice to log all retries - then we could check if a retried query could have caused the problem.

@psarna the --max-mutation-retries  is set to 5 for all gemini tests

**Comment by psarna** [comment]
I ran the test again and confirmed with extended logs that the duplicated values are a direct result of a retried operation on a list, so it's #3559.

In short, the issue goes like that - overwriting an unfrozen list with timestamp `t` is implemented in Scylla as two mutations:
1. create a tombstone for each list cell with timestamp `t-1`
2. create a new value with timestamp `t`

Now, if this operation is retried and as a result sent to more than one node, all of the involved nodes would create such mutations. Unfortunately, after reconciliation the order wrt. user provided timestamps is as follows:
1. ts=`t-1`: tombstone created on node A
2. ts=`t-1`: tombstone created on node B
3. ts=`t`   : value created on node A
4. ts=`t`   : value created on node B

... hence the duplicated values. It's a "feature" of CQL lists, so technically not an error, but expected behavior.
Gemini should avoid overwriting lists with user-provided timestamps **and** allow retrying this operation, because this operation is known to not be idempotent.

**Comment by eliransin** [comment]
> ... hence the duplicated values. It's a "feature" of CQL lists, so technically not an error, but expected behavior.
> Gemini should avoid overwriting lists with user-provided timestamps **and** allow retrying this operation, because this operation is known to not be idempotent.

@penberg I assigned this to you since AFAICR you are the father of Gemini :slightly_smiling_face: , please reassign if someone else should handle this.

**Comment by nuivall** [comment]
Yes, understand now. I think it happens because it implements retry with the same timestamp.

**Comment by nyh** [comment]
> It's possible that Gemini generated two queries that insert a list with the same timestamp. In such case there will be two values in the list, even on a single node.
> 
> ```sql
> cqlsh:ks> CREATE TABLE t (p int primary key, l list<int>);
> cqlsh:ks> INSERT INTO t (p, l) VALUES (0, [1]) USING TIMESTAMP 1337;
> cqlsh:ks> INSERT INTO t (p, l) VALUES (0, [1]) USING TIMESTAMP 1337;
> cqlsh:ks> SELECT * FROM t;
> 
>  p | l
> ---+--------
>  0 | [1, 1]
> 
> (1 rows)
> ```

Indeed. This is exactly issue #3559 (as Piotr Sarna noticed above).

Maybe  we should close this issue here (as a duplicate of #3559), and if we want to fix Gemini not to fall on this known "bug" (or arguably, a not-very-desirable "feature"), we should open a Gemini issue.

**Comment by nuivall** [comment]
closing, gemini was fixed in https://github.com/scylladb/gemini/pull/373/commits/fd281582bcae7d2d2f052e66bf207115836350d2
